RUMM-1667 Enhance release automation

ncreated opened the pull request:
### What and why?

📦 This PR proposes enhancement to our release automation model. Compared to previous model:
- It no longer creates `Datadog-x.y.z.zip` archive in current directory, instead it clones the repo to temporary folder which ensures clean state for building `XCFrameworks`. It removes the need for [`prepare()` and `cleanup()`](https://github.com/DataDog/dd-sdk-ios/blob/920464537fd09d0b7166ff030fae9040138aab01/tools/distribution/make_distro_builds.sh#L26-L42) operations and guarantees that no dirty artefacts will be included in the `.zip` (in `1.7.0`, `1.7.1` and `1.8.0-beta1` we accidentally included `HTTPServerMock.xcframework`).
- It runs integrity checks to ensure that [`sdk_version`](https://github.com/DataDog/dd-sdk-ios/blob/920464537fd09d0b7166ff030fae9040138aab01/Sources/Datadog/Versioning.swift#L3) and [`spec.version`](https://github.com/DataDog/dd-sdk-ios/blob/920464537fd09d0b7166ff030fae9040138aab01/DatadogSDK.podspec#L4) match the git tag and validates the content of produced `Datadog-x.y.z.zip` archive.
- It supports `--dry-run` mode for testing and debugging locally and on CI (with `DD_RELEASE_DRY_RUN=1` ENV).
- It allows selective run for particular release tags (also on CI) which will be necessary for fixing `1.7.0`, `1.7.1` and `1.8.0-beta1` GH assets.

### How?

#### Updated Bitrise pipelnie

I reshaped our Bitrise CI pipeline to spin off two distinct jobs, one for publishing GH Release asset, one for pushing pods:
```
tagged_commit:               --- (tests) ---> ✅
publish_github_asset:                       ↘ --(make .zip + validate + upload)--> ✅
publish_cocoapods_podspecs:                 ↘ --(lint specs + push specs)-----------------------> ✅
```
I used the official [Bitrise Start Build](https://www.bitrise.io/integrations/steps/build-router-start) step instead of calling Bitrise API with `curl` (it gives us ENV propagation from `tagged_commit` to child jobs for free). It is possible to support Github Check by making the `tagged_commit` await child jobs completion, but in our case this will always end up with a CI timeout (`90min`) as Cocoapods specs push might take `~1h`. Instead, eventual failure from child jobs is reported to our Slack channel.

#### New CLI tooling

I replaced Bash scripts with more flexible Python CLI tool:
```
❯ ./tools/distribution/release.py -h
usage: release.py [-h] [--only-github] [--only-cocoapods] [--overwrite-github] [--dry-run] git_tag

positional arguments:
  git_tag             Git tag name.

optional arguments:
  -h, --help          show this help message and exit
  --only-github       Only publish GH Release asset.
  --only-cocoapods    Only publish Cocoapods podspecs.
  --overwrite-github  Overwrite existing GH Release asset.
  --dry-run           Run validation but skip publishing.
```
It uses `python3` and no `pip` dependencies. The `--dry-run` option allows for local and remote (`DD_RELEASE_DRY_RUN=1`) debugging. All options have their ENV counterpart, meaning that it can be run for any release tag locally and on CI.

The tool itself is simple: it makes, validates and publishes GH Release asset and / or Cocoapod specs, e.g.:
```python
gh_asset = GHAsset()  # make the `.zip`
gh_asset.validate(git_tag=git_tag)  # validate it
gh_asset.publish(git_tag=git_tag, overwrite_existing=overwrite_github, dry_run=dry_run)  # upload to GH
```
Example CI logs for `--dry-run`: [publishing GH Asset](https://app.bitrise.io/build/7d4e881d-7b8f-4661-9d6b-d812c872d5b8#?tab=log), [publishing CP specs](https://app.bitrise.io/build/f7ea17f8-262d-4f3c-b056-de862f6ece8b#?tab=log).

### Review checklist

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference

ncreated commented on tools/distribution/src/assets/podspec.py:
```diff
@@ -0,0 +1,107 @@
+        # but this is also limited by the CI job timeout.
+        retry_time = 30  # seconds
+        attempt = 0
+        while True or attempt > 100:  # as trunk may not yet
```
```suggestion
        while True or attempt > 100: 
```

ncreated commented on tools/distribution/src/assets/podspec.py:
```diff
@@ -0,0 +1,107 @@
+        """
+        Reads pod version from podspec file.
+        """
+        version_regex = r'^.*\.version.*=.*\"([0-9]\.[0-9]\.[0-9][\-a-z0-9]*)\"'  # e.g. ' s.version   = "1.7.1-alpha1"'
```
```suggestion
        version_regex = r'^.*\.version.*=.*\"([0-9]+\.[0-9]+\.[0-9]+[\-a-z0-9]*)\"'  # e.g. ' s.version   = "1.7.1-alpha1"'
```
Otherwise, it will fail for `1.10.0`.

maxep commented on tools/distribution/src/assets/podspec.py:
```diff
@@ -0,0 +1,107 @@
+                    shell(f'pod trunk push --synchronous --allow-warnings {self.__file_name}', skip=dry_run)
+                else:
+                    shell(f'pod repo update --silent', skip=dry_run)
+                    shell(f'pod spec lint  --silent --allow-warnings {self.__file_name}', skip=dry_run)
+                    shell(f'pod trunk push --allow-warnings {self.__file_name}', skip=dry_run)
```
I *think* `pod trunk push` performs the same validation as  `pod spec lint`.

ncreated replied:
Well, actually we can have both by adding a bit more complexity to this tool and capturing `shell('')` output along with its result. Then we will control which output to print in the log. So this is a third option to consider 👈 .

ncreated commented:
I added an improvement to always build our dependencies from source, which leads to producing minimal GH Asset (with only the architecture slices we need):
```diff
- shell('carthage bootstrap --no-build')
- shell('carthage build --platform iOS --use-xcframeworks --no-skip-current')
+ shell('carthage bootstrap --platform iOS --no-build')
+ shell('carthage build --platform iOS --use-xcframeworks --no-use-binaries --no-skip-current')
```
Also, added additional log on the `XCFrameworks` details. I tried with `1.7.1` and `--dry-run`:
```
→ details on bundled `XCFrameworks`:
   - /Kronos.xcframework/ios-arm64_i386_x86_64-simulator
   - /Kronos.xcframework/Info.plist
   - /Kronos.xcframework/ios-arm64_armv7
   - /DatadogCrashReporting.xcframework/ios-x86_64-simulator
   - /DatadogCrashReporting.xcframework/ios-arm64
   - /DatadogCrashReporting.xcframework/Info.plist
   - /DatadogObjc.xcframework/ios-arm64_x86_64-simulator
   - /DatadogObjc.xcframework/ios-arm64
   - /DatadogObjc.xcframework/Info.plist
   - /Datadog.xcframework/ios-arm64_x86_64-simulator
   - /Datadog.xcframework/ios-arm64
   - /Datadog.xcframework/Info.plist
   - /CrashReporter.xcframework/ios-arm64_i386_x86_64-simulator
   - /CrashReporter.xcframework/ios-arm64_arm64e_armv7_armv7s
   - /CrashReporter.xcframework/Info.plist
```
